File: claude_exp/routers/router_py/shared/log_throttle.py
import logging
import threading


class ThrottledLogger:
    """Wraps a logger for conditions that can fire once per transaction (crypto_host call
    failures, dropped 0110s) - at soak TPS, plus the deliberate fail-percentage/chaos scenarios
    in briefs/resilience_v2.md, these would otherwise flood stdout badly enough to make a stress
    run unreadable (mirrors upstream_host/upstream_shared/log_throttle.py's own reasoning for
    0120/0420 advice traffic - same flood, this side of the router).

    Logs the first occurrence of a condition in full, then only every `every`-th repeat after
    that (with the running count appended) - so the condition is never silent, but volume drops
    by ~`every`x. Counts are per `key`, not per call site, so unrelated conditions throttle
    independently.

    Diverges from upstream_host's copy by accepting `extra` and passing it through: dispatcher's
    call sites rely on extra={"router_stan": ...} so json_log.py can still join a throttled line
    back to the transaction it belongs to (see _EXTRA_FIELDS)."""

    def __init__(self, logger: logging.Logger, every: int = 200):
        self._logger = logger
        self._every = every
        self._counts = {}
        self._lock = threading.Lock()

    def log(self, level: int, key: str, fmt: str, *args, extra: dict = None) -> None:
        with self._lock:
            count = self._counts[key] = self._counts.get(key, 0) + 1
        if count == 1:
            self._logger.log(level, fmt, *args, extra=extra)
        elif count % self._every == 0:
            self._logger.log(
                level, fmt + " [occurrence #%d of this condition, throttled]", *args, count, extra=extra
            )

File: claude_exp/routers/router_java/src/main/java/com/router/shared/LogThrottle.java
package com.router.shared;

import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;
import java.util.logging.Level;
import java.util.logging.Logger;

/** Wraps a Logger for conditions that can fire once per transaction (crypto_host call failures,
 * dropped 0110s) - briefs/resilience_v2.md's fail_percentage/crypto-kill chaos scenarios turn a
 * rare warning into a sustained per-transaction flood without this (mirrors router_py's
 * shared/log_throttle.py::ThrottledLogger 1:1).
 *
 * Logs the first occurrence of a condition in full, then only every `every`-th repeat after that
 * (with the running count appended) - so the condition is never silent, but volume drops by
 * ~`every`x. Counts are per key, not per call site, so unrelated conditions throttle
 * independently. */
public final class LogThrottle {
    private final Logger logger;
    private final int every;
    private final ConcurrentHashMap<String, AtomicLong> counts = new ConcurrentHashMap<>();

    public LogThrottle(Logger logger, int every) {
        this.logger = logger;
        this.every = every;
    }

    public void log(Level level, String key, String message) {
        long count = counts.computeIfAbsent(key, k -> new AtomicLong(0)).incrementAndGet();
        if (count == 1) {
            logger.log(level, message);
        } else if (count % every == 0) {
            logger.log(level, message + " [occurrence #" + count + " of this condition, throttled]");
        }
    }
}
